core/src/main/scala/com/eed3si9n/jarjarabrams/Zip.scala

```diff
@@ -1,192 +1,131 @@
-            import Attributes.Name.MANIFEST_VERSION
-            val main = mf.getMainAttributes
-            if (!main.containsKey(MANIFEST_VERSION))
-              main.put(MANIFEST_VERSION, "1.0")
-            val os = new JarOutputStream(fileOut)
-            val e = new ZipEntry(JarFile.MANIFEST_NAME)
-            e.setTime(time)
-            os.putNextEntry(e)
-            mf.write(new BufferedOutputStream(os))
-            os.closeEntry()
-            (os, "jar")
-          case None => (new ZipOutputStream(fileOut), "zip")
-        }
-      try {
-        f(zipOut)
-      } finally {
-        zipOut.close
-      }
-    }
-  }
-
-  private def unzipStream(
-      from: InputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean = false
-  ): Set[Path] = {
-    createDirectories(toDirectory)
-    Using.zipInputStream(from)(zipInput => extract(zipInput, toDirectory, preserveLastModified))
-  }
-
-  private def extract(
-      from: ZipInputStream,
-      toDirectory: Path,
-      preserveLastModified: Boolean
-  ) = {
-    val set = new mutable.HashSet[Path]
-    @tailrec def next(): Unit = {
-      val entry = from.getNextEntry
-      if (entry == null) ()
-      else {
-        val name = entry.getName
-        val target = toDirectory.resolve(name)
-        if (entry.isDirectory) createDirectories(target)
-        else {
-          set += target
-          try {
-            Using.fileOutputStream(false)(target)(out => transfer(from, out))
-          } catch {
-            case NonFatal(e) =>
-              throw new RuntimeException(
-                "error extracting Zip entry '" + name + "' to '" + target + "'",
-                e
-              )
+  /**
+   * Treating JAR file as a set of EntryStruct, this implements a
+   * functional processing, intended for in-memory shading.
+   */
+  def transformJarFile(
+      inputJar: Path,
+      outputJar: Path,
+      resetTimestamp: Boolean
+  )(f: EntryStruct => Option[EntryStruct]): Path =
+    Using.jarFile(inputJar) { in =>
+      val tempJar = Files.createTempFile("jarjar", ".jar")
+      Using.jarOutputStream(tempJar) { out =>
+        val names = new mutable.HashSet[String]
+        in.entries.asScala.toList.foreach { entry0 =>
+          val struct0 = entryStruct(
+            entry0.getName,
+            entry0.getTime,
+            toByteArray(in.getInputStream(entry0)),
+            skipTransform = false
+          )
+          f(struct0) match {
+            case Some(struct) =>
+              if (names.add(struct.name)) {
+                val entry = new JarEntry(struct.name)
+                val time =
+                  if (resetTimestamp) hardcodedZipTimestamp(struct.name)
+                  else enforceMinimum(struct.time)
+                entry.setTime(time)
+                entry.setCompressedSize(-1)
+                out.putNextEntry(entry)
+                out.write(struct.data)
+              } else if (struct.name.endsWith("/")) ()
```

Comment:
Should this not be first? Not sure why it is here actually. The else of this if is also returning ()

Reply:
I think we do want entries that ends with `/` to transform like `shapeless/` becomes `foo/shapeless`. The original `else if` in Java code exempts directories from duplicate entry exception. If we ignore the duplicate entry exception, like I am doing here, then this clause is meaningless.